src/org/sosy_lab/java_smt/solvers/cvc5/CVC5InterpolatingProver.java

```diff
@@ -0,0 +1,334 @@
+    Preconditions.checkArgument(
+        !partitionedFormulas.isEmpty(), "at least one partition should be available.");
+
+    final List<BooleanFormula> itps = new ArrayList<>();
+    for (int i = 1; i < partitionedFormulas.size(); i++) {
+      itps.add(
+          getInterpolant(
+              ImmutableList.copyOf(Iterables.concat(partitionedFormulas.subList(0, i)))));
+    }
+    return itps;
+  }
+
+  // Experimental!
+  @Override
+  public List<BooleanFormula> getTreeInterpolants(
+      List<? extends Collection<Term>> pPartitionedFormulas, int[] pStartOfSubTree)
+      throws SolverException, InterruptedException {
+    Preconditions.checkState(!closed);
+    assert InterpolatingProverEnvironment.checkTreeStructure(
+        pPartitionedFormulas.size(), pStartOfSubTree);
+
+    // Generate every Interpolation Pair
+    ArrayList<ArrayList<ArrayList<Collection<Term>>>> interpolationPairs =
+        getTreeInterpolationPairs(pPartitionedFormulas, pStartOfSubTree);
+
+    final ArrayList<Term> itps = new ArrayList<>();
+    try { // Interpolate every Interpolation Pair
+      for (int i = 0; i < interpolationPairs.size(); i++) {
+        itps.add(getCVC5Interpolation(interpolationPairs.get(i)));
+      }
+    } catch (UnsupportedOperationException e) {
+      if (e.getMessage() != null) {
+        throw new SolverException(e.getMessage(), e);
+      } else {
+        throw e;
+      }
+    }
+
+    final List<BooleanFormula> result = new ArrayList<>();
+    for (Term itp : itps) {
+      result.add(creator.encapsulateBoolean(itp));
+    }
+    assert result.size() == pStartOfSubTree.length - 1;
+    return result;
+  }
+
+  /**
+   * Interpolates a Tuple of Interpolants according to Craig-Interpolation using CVC5-Interpolation.
+   *
+   * @param formulaPair Interpolation Pair
+   * @return Interpolation of the Interpolation Pair following the definition of
+   *     Craig-Interpolation.
+   */
+  private Term getCVC5Interpolation(ArrayList<ArrayList<Collection<Term>>> formulaPair) {
+    assert formulaPair.size() == 2; // Check that formulaPair is a Tuple
+
+    ArrayList<Collection<Term>> assertedInterpols = formulaPair.get(0);
+    ArrayList<Collection<Term>> addedInterpols = formulaPair.get(1);
+
+    // Respect Asserted Formulas not in the Interpolation Pairs
+    ArrayList<Collection<Term>> combinedInterpols = new ArrayList<>();
+    combinedInterpols.addAll(assertedInterpols);
+    combinedInterpols.addAll(addedInterpols);
+
+    Collection<Term> extraAssertions = getAssertedTermsNotInCollection(combinedInterpols);
+    Term extraAssert = new Term();
+
+    if (extraAssertions.isEmpty()) {
+      extraAssert = solver.mkTrue();
+    } else {
+      extraAssert = buildConjunctionOfFormulas(extraAssertions);
+    }
+
+    Term phim = buildConjunctionOfFormulasOverArray(assertedInterpols);
+    Term phip = buildConjunctionOfFormulasOverArray(addedInterpols);
+
+    solver.resetAssertions();
```

Comment:
Just asking:
Is this the solver of this ProverEnvironment-instance?
We must delete the entire solver's context to solve interpolation?
So...ah... how do you rebuild the context after interpolation is done?

Reply:
Could we use a separate solver-instancee for interpolation and keep the original solver-context unmodified?

Reply:
Yes, the newest commit uses a separate solver-instance and leaves to original solver-context untouched.

Reply:
Just for clarification, let me ask some questions.

Technical part:
- Does CVC5 allow to share formulas (Terms) between distinct Solver instances?  
JavaSMT already has the possibility for several ProverEnvironments for cvC5, so that part should alredy work as expected, without the requirement of translating formulas or synchronizing symbol declarations. Correct?
- Do you need the solver instance only for the interpolation query itself or for the full lifetime of the Proverenvironment? Is there any information kept in the Solver after returning from the interpolation query?
If not, please do not create the Solver-instance on class-level, but only as part of the interpolation-method.
And check whether it can be closed afterwards.

Theoretical part:
- Can you describe/document how the interpolation based on the CVC5-interface should work? 
- Until now, I could not see any check on symbol usage in the interpolant. Does a CVC5-provided interpolant only contain symbols that are shared between both input formulas? Is this specified in the CVC5-API?

Deep theoretical part (for later documentation): :-)
- Where is the guarantee that a sequence of interpolants satisfies the interpolation criteria of sequntial interpolation?

Reply:
Technical part: 
 - Sharing formulas between different solver instances works.
 - I changed it to exist only during the interpolation-method. One thing would be kept in the solver. If the interpolation call was successful, you could call `getInterpolantNext` directly after to produce a different but also valid interpolant. In the case the assertions do not change, i could "redirect" a second interpolation call to this method. Would this be wished behavior? 
 
The rest is discussed below.

Reply:
There is currently no requirement for computing more than one interpolant (or one set of sequential/tree interpolants) per SAT-check. So it can be kept simple and without any state-management.